### ISSUE TITLE
[Ch 04][Notebook][Major] "두 방식을 나란히 학습" 이 실제로는 동일한 코드 2벌 — 동등성이 코드로 시연되지 않음

### ISSUE THREAD

**Issue by modesta-jihye-lee** [yoon-gu/neuqes-101#97]
## 대상

- **챕터**: Ch 04 — Binary on softmax (CrossEntropy 등장 + sigmoid+BCE와의 동등성)
- **산출물**: Notebook (`04_softmax_binary/04_softmax_binary.ipynb` — 실습 표 `cell009`, 학습 셀 `cell010`, 변형 셀 `cell015`, 체크포인트 `cell017`)
- **기준 커밋**: `eed3dbe`
- **환경**: 정적 검토 + 실행본 `executed/04_softmax_binary.ipynb` 대조 + 로컬 재실행(sklearn 1.9.0) 에러 0건 / 11초. **Colab 전체 실행 완료** (2026-08-15 · CPU 런타임 · 45초 · 에러 0건)

## 실제 결과

이 챕터의 표제 주장은 "**두 방식을 나란히 학습해 결과가 같음을 확인**" 입니다.

- 챕터 목표(`cell001`): "두 방식이 수학적으로 동등하다는 것을 **식과 코드로 직접 확인**"
- 학습 흐름 1: "같은 Yelp 이진화 데이터에 **두 방식을 학습** — sigmoid+BCE(Ch 3 그대로) vs softmax+CE(이번 챕터)"
- `cell009` 표:

  | 방식 | sklearn 인자 | 출력 차원 | 활성화 | loss |
  |---|---|---|---|---|
  | A (Ch 3 그대로) | `LogisticRegression()` | 1 | sigmoid | BCE |
  | B (이번 챕터) | `LogisticRegression()` (multinomial 자동) | 2 | softmax | CE |

그런데 `cell010` 의 두 모델은 **문자 그대로 같은 코드**입니다.

```python
model_a = LogisticRegression(max_iter=1000)   # 방식 A
model_a.fit(X_train, y_train)

model_b = LogisticRegression(max_iter=1000)   # 방식 B — 인자가 A 와 완전히 동일
model_b.fit(X_train, y_train)
```

같은 모델·같은 설정·같은 데이터이므로, 결과가 같은 것은 **두 방식이 동등하다는 증거가 아니라 같은 코드를 두 번 돌리면 같은 답이 나온다는 사실**일 뿐입니다. 출력의 `Diff: 0.0000`, `Max diff: 0.0000`, `coef_ max diff: 0.00e+00` 은 두 방식이 같아서가 아니라 같은 모델이라서 나온 값입니다.

### 그래서 서술과 출력이 어긋납니다

챕터는 세 곳에서 "미세한 차이가 있지만 사실상 같다" 로 서술하는데, 실제 출력은 **정확히 0** 입니다.

| 위치 | 서술 | 실제 출력 |
|---|---|---|
| `cell015` print 문 | `(small difference is only solver convergence noise; same model essentially)` | `coef_ max diff: 0.00e+00` · `intercept_ max diff: 0.00e+00` |
| `cell014` 본문 | "두 방식이 그래서 사실상 같은 모델이 되어 `predict_proba` 도 **거의** 일치하는 거였죠" | `Max diff: 0.0000` (완전 일치) |
| FAQ Q4 | "`predict_proba` 도 **미세한 수치 오차 빼고** 일치합니다" | 동일 |

같은 모델을 두 번 학습시킨 것이므로, 원문이 말하는 "solver convergence noise"(계산 과정에서 생기는 미세한 오차) 는 애초에 생길 수 없습니다.

### 체크포인트 질문 3번이 노트북 내용으로 답할 수 없습니다

> 3. **방식 B의 두 coefficient 벡터** 사이에 어떤 관계가 학습되는 경향이 있나요? 그 이유는?

이 챕터에서 방식 B 의 `coef_` 는 `(1, 10000)` 입니다 — **두 coefficient 벡터가 아예 존재하지 않습니다**(그게 이 챕터 변형 절의 결론이기도 합니다). 학습자가 노트북 안에서 답을 찾을 근거가 없습니다.

## 기대 결과

챕터의 설명(sklearn 은 클래스가 2개면 2차원 출력을 만들지 않고 1차원으로 축약한다)은 `cell014`·FAQ Q4·Q5 에 이미 정확히 적혀 있습니다. **문제는 실습 셀이 그 사실보다 앞에 놓여 "두 설정을 비교한 것처럼" 보인다는 것**입니다. 다음 중 하나로 정리하는 것을 제안합니다.

**(안 1) 정직한 순서로 재배치 — 코드 변경 최소**

`cell009` 표와 `cell010` 주석에서 먼저 밝힙니다.

> sklearn 에서는 이진 데이터에 두 방식을 *따로* 학습시킬 수 없습니다 — 클래스가 2개면 `LogisticRegression()` 이 출력을 2차원으로 만들지 않고 **자동으로 1차원으로 축약**하기 때문입니다(FAQ Q4). 그래서 아래 두 줄은 사실 **같은 코드**이고, 결과가 같은 것은 당연합니다. 두 방식이 정말 같은지는 이어지는 §해부 에서 *식과 numpy* 로 직접 확인합니다.

그리고 `cell015` 의 `(small difference is only solver convergence noise...)` 문구를 `(exactly identical — 같은 모델을 두 번 학습시킨 것이므로 당연히 0)` 으로 교체합니다.

**(안 2) 진짜 방식 B 를 numpy 로 한 번 학습** — 2차원 logit + softmax + CE 를 직접 구현해 `coef_` 두 벡터가 $w_1 - w_0$ 만 의미 있음을 보이면, 체크포인트 3번도 답할 수 있게 됩니다.

어느 쪽이든 `cell013`(임의 logit 쌍으로 $\text{softmax}([z_0,z_1])_1 = \sigma(z_1-z_0)$ 검증)은 이 챕터에서 **실제로 동등성을 보여 주는 유일한 코드**이므로 그대로 두고 비중을 키우는 편이 좋겠습니다.

## 재현 절차

```bash
python3 -c "
import json
nb = json.load(open('04_softmax_binary/04_softmax_binary.ipynb'))
for c in nb['cells']:
    if c.get('id') == 'cell010': print(''.join(c['source']))
"
```

두 `LogisticRegression(max_iter=1000)` 호출의 인자가 동일한지 눈으로 확인하면 됩니다.

## 로그

```
Method A (sigmoid + BCE) accuracy: 0.8639
Method B (softmax + CE)  accuracy: 0.8639
Diff: 0.0000

Method A coef_ shape:      (1, 10000)
Method B coef_ shape:      (1, 10000)
coef_ max diff:      0.00e+00      ← "solver convergence noise" 가 아니라 정확히 0
intercept_ max diff: 0.00e+00

(small difference is only solver convergence noise; same model essentially)   ← 출력 스스로 모순
```

**Colab 재실행(2026-08-15)에서도 위 값이 전부 동일했습니다.** Ch 1·2 에서 나타난 환경별 수치 드리프트가 이 챕터에는 없었고, `coef_ max diff` 는 양쪽 실행 모두 정확히 `0.00e+00` 입니다 — 즉 "미세한 차이" 서술은 어느 환경에서도 성립하지 않습니다.

## 전파 범위

| 산출물 | 상태 |
|---|---|
| `04_softmax_binary/04_softmax_binary.ipynb` `cell009`·`cell010`·`cell015`·`cell017` | ❌ |
| `pages/04-softmax_binary-practice.md`, `-variation.md`, `-wrapup.md` | ❌ (같은 코드·서술) |
| `book/chapters/ch04_softmax_binary.tex` | ❌ → EPUB 영향 |

## 심각도

**Major** — CLAUDE.md 가 규정한 이 챕터의 존재 이유("동일 binary 데이터에 두 방식을 학습해 `predict_proba` 가 일치함을 시연")가 코드 수준에서 성립하지 않고, 출력과 서술("계산 오차 때문에 미세하게 다르다")이 어긋납니다. 실행은 정상이므로 Blocker 는 아닙니다.

## 재실행 여부

**(안 1)** 이면 마크다운·주석·print 문 수정이라 재실행이 필요하지만 CPU 챕터라 부담이 작습니다(11초). **(안 2)** 는 셀 추가이므로 재실행 필수. 이후 `executed/` 갱신 → WikiDocs 재변환 → EPUB 재빌드가 필요합니다.

## 담당

- 수정 담당자: 미정
- 재검수자: 등록자 (@modesta-jihye-lee)
- 연결: 챕터 검수 이슈 #28
